cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -119,7 +125,13 @@ struct GrouperImpl : Grouper {
     }
 
     for (int i = 0; i < batch.num_values(); ++i) {
-      RETURN_NOT_OK(encoders_[i]->Encode(batch[i], batch.length, key_buf_ptrs.data()));
+      ExecValue value;
+      if (batch[i].is_array()) {
+        value.SetArray(*batch[i].array());
+      } else {
+        value.SetScalar(batch[i].scalar().get());
+      }
+      RETURN_NOT_OK(encoders_[i]->Encode(value, batch.length, key_buf_ptrs.data()));
```

Comment:
This is temporary because Grouper and things hashing related are going to switch soon to use ExecSpan, so this will not be necessary once that's done